src/Servers/HttpSys/src/RequestProcessing/TlsListener.cs

```diff
@@ -0,0 +1,95 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Server.HttpSys.RequestProcessing;
+
+internal sealed partial class TlsListener : IDisposable
+{
+    private readonly ConcurrentDictionary<ulong, DateTime> _connectionTimestamps = new();
+    private readonly Action<IFeatureCollection, ReadOnlySpan<byte>> _tlsClientHelloBytesCallback;
+
+    private readonly ILogger _logger;
+    private readonly CancellationTokenSource _cts = new();
+    private readonly Task _cleanupTask;
+
+    private static readonly TimeSpan ConnectionIdleTime = TimeSpan.FromMinutes(10);
+    private static readonly TimeSpan CleanupInterval = TimeSpan.FromSeconds(30);
+
+    internal TlsListener(ILogger logger, Action<IFeatureCollection, ReadOnlySpan<byte>> tlsClientHelloBytesCallback)
+    {
+        _logger = logger;
+        _tlsClientHelloBytesCallback = tlsClientHelloBytesCallback;
+
+        _cleanupTask = Task.Run(() => CleanupLoopAsync(_cts.Token));
+    }
+
+    internal void InvokeTlsClientHelloCallback(IFeatureCollection features, Request request)
+    {
+        if (!request.IsHttps)
+        {
+            return;
+        }
+
+        if (_connectionTimestamps.ContainsKey(request.RawConnectionId))
```

Comment:
For Http/2 this could result in the callback being called multiple times for parallel requests on connection startup.

`_connectionTimestamps.TryAdd` might be able to fix this race.

Reply:
I want to add only if tls client hello callback successfully worked, so I would like to leave it here like that. Do you know how can i enforce a single run of tls client hello callback fetch even if there are multiple parallel requests?

Reply:
updated to `TryAdd`